### PR TITLE
Allow accessing CORS routes on ApiControllers if a proper CSRF token is provided

### PR DESCRIPTION
**Pull request by juliusknorr**
The current implementation always requires basic auth when `@CORS` annotated routes are accessed. That leads to apps implementing basically two endpoints that have the same implementation but one with `@CORS @NoCSRFRequired` and one without those.

This PR would basically allow apps to provide one API endpoint for both internal usage in Nextcloud as well as external API calls, but the CORSMiddleware doesn't block the request if a valid CSRF token is provided.

I think this should cover the CSRF attack vectors that are described in https://github.com/nextcloud/server/blob/b801d2765868c90580e8f1097e225f590db62003/lib/private/AppFramework/Middleware/Security/CORSMiddleware.php#L85-L86 or am I missing something?

---

The second commit is just for checking if username and password are actually provided so that the type hint added in https://github.com/nextcloud/server/blob/1b46621cd31dabb48084c8e23f741de77e7007b4/lib/public/User/Events/BeforeUserLoggedInEvent.php#L46 doesn't cause a `TypeError` to be thrown but when calling `logClientIn` in the CORSMiddleware.

### CODE REVIEW

File: lib/private/AppFramework/Middleware/Security/CORSMiddleware.php [hunk 1]
```diff
@@ -26,6 +26,7 @@
 
 namespace OC\AppFramework\Middleware\Security;
 
+use OC\AppFramework\Middleware\Security\Exceptions\CrossSiteRequestForgeryException;
```
Comment (ChristophWurst) [inline lib/private/AppFramework/Middleware/Security/CORSMiddleware.php]: nit: not used?

Comment (juliushaertl) [reply lib/private/AppFramework/Middleware/Security/CORSMiddleware.php]: Ah, not that was a leftover from my first attempt to implement everything in the CORS middleware

**Comment by nickvergessen** [comment]
And just using the OCS api internally as well does not work because?

**Comment by juliushaertl** [comment]
> And just using the OCS api internally as well does not work because?

Yes that would also work of course. But this is not documented anywhere, while the app tutorial explains how the CORS routes work. Personally I also don't see any benefit in wrapping my apps API routes in to the OCS wrapper - is there any? 

If we want to encourage app developers to use OCS then we need to explain that better.

**Comment by nickvergessen** [comment]
> is there any?

Well it makes sure thet stuff works with mobile apps for free :)

That is exactly why we built OCS and OCS-Resources even. In Talk the full API is OCS. I think this should just be the way to go instead.

**Comment by juliushaertl** [comment]
> That is exactly why we built OCS and OCS-Resources even. In Talk the full API is OCS. I think this should just be the way to go instead.

Could you provide some more insights on the decision back then? As to me the current way that API access is limited to OCS endpoints feels more like an artificial barrier, so app developers are pinned to the way OCS responses are structured.

Just one thing I encountered as of today is that with the current way OCSResponses are implemented it is not possible to emit a proper OCS response in a middleware that catches exceptions in order to convert them into responses. I use that approach for [catching exceptions during controller execution in deck](https://github.com/nextcloud/deck/blob/master/lib/Middleware/ExceptionMiddleware.php#L64-L72) so the API returns a standardized error response and there is less duplicated code for the exception handling in the various controller methods.

**Comment by rullzer** [comment]
> > That is exactly why we built OCS and OCS-Resources even. In Talk the full API is OCS. I think this should just be the way to go instead.
> 
> Could you provide some more insights on the decision back then? As to me the current way that API access is limited to OCS endpoints feels more like an artificial barrier, so app developers are pinned to the way OCS responses are structured.
> 
> Just one thing I encountered as of today is that with the current way OCSResponses are implemented it is not possible to emit a proper OCS response in a middleware that catches exceptions in order to convert them into responses. I use that approach for [catching exceptions during controller execution in deck](https://github.com/nextcloud/deck/blob/master/lib/Middleware/ExceptionMiddleware.php#L64-L72) so the API returns a standardized error response and there is less duplicated code for the exception handling in the various controller methods.

Then we'd need to fix that. However the OCS endpoints are public by definition. So i would propose to first fix cors there.

**Comment by juliushaertl** [comment]
Ok, so I guess we don't do that then.

**Comment by juliushaertl** [comment]
> Could you provide some more insights on the decision back then? As to me the current way that API access is limited to OCS endpoints feels more like an artificial barrier, so app developers are pinned to the way OCS responses are structured.

I'd still be interested in some reasoning about the OCS enforcement, I'd be happy to make that clearer in our developer docs then.

**Pull request closed without merging**